### PR TITLE
Better codegen quality when emitting degenerate switches.

### PR DESCRIPTION
**Pull request by VSadov**
Related to dotnet/coreclr#7914
Where the offending pattern was observed in the IL generated for the Kestrel web server and was recommended as an easy improvement.

Because of cascaded dispatching into "try" regions , it is not uncommon to see async state machine to contain degenerate switches like

IL_0008:  switch     (
                          IL_003b,
                          IL_003b,
                          IL_003b,
                          IL_003b,
                          IL_003b,
                          IL_003b,
                          IL_003b,
                          IL_0729)

Note that numerous cases all get dispatched to the same target regardless of the actual value.
We can trivially emit such switches as just a range-check (which switch would need to do anyways), without any actual "switching".
Since all cases lead to the same label, there is no need to switch once we know the value is in range of the switch.

Fixes:#14878

**Customer scenario**

Improves codegen quality in async methods containing exception handling (turns out it is common).

**Bugs this fixes:** 

#14878

**Workarounds, if any**

There is no workaround in the case of synthesized code since user has no direct access to it.
It is however just an improvement, not strictly a fix.
 Current way of emitting is correct, just not the best.

**Risk**

The change is specific to the way a certain form of switches is emitted.

**Performance impact**

Low impact at compile time.
Slight improvement of the generated code.

**Is this a regression from a previous update?**

NO

**Root cause analysis:**

We did not realize that degenerate switches are common enough in the synthesized code.

**How was the bug found?**

JIT team was investigating ways to improve perf of Kestrel web server and listed better codegen for degenerate switches as one of the actionable items.

### CODE REVIEW

**Comment by VSadov** [comment]
@jaredpar - Not sure if this meets the bar for the master. Could as well go into the future branch.

**Comment by jaredpar** [comment]
If this is affecting Kestrel web server performance then I think it meets the bar.  That's a core scenario for us.

**Comment by VSadov** [comment]
@dotnet-bot test windows_determinism_prtest please

Seems like a network issue while processing packages.

**Comment by jaredpar** [comment]
retest determinism please

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - please review

File: src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs [hunk 1]
```diff
@@ -1092,7 +1096,22 @@ private void EmitSwitchStatement(BoundSwitchStatement switchStatement)
                         var value = boundLabel.ConstantValueOpt;
                         Debug.Assert(value != null
                             && SwitchConstantValueHelper.IsValidSwitchCaseLabelConstant(value));
-                        labelsBuilder.Add(new KeyValuePair<ConstantValue, object>(value, label));
+
+                        // in debug we should be able to set breakpoints on individual case lables,
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs]: I don't know what this comment refers to. We do not support breakpoints on case labels, and there is no `nop` instruction between them as there would need to be in order to do so.

File: src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs [hunk 2]
```diff
@@ -60,28 +82,88 @@ internal ulong BucketSize
                 }
             }
 
-            // Relative cost of the bucket
-            // roughly proportional to the number of compares it needs in the success case.
-            internal int BucketCost
+            // if a bucket could be split into two degenerate ones
+            // specifies a label index where that split would occur
```
Comment (gafter) [inline src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs]: When you say "where that split would occur", does it occur *before* that index, *after* that index, or both?

Comment (VSadov) [inline src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs]: The index would be the start of the second bucket. That is why -1 and 0 can have special meaning.
Will modify the comment.

File: src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs (hunk 1 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs]: We probably should. We do allow breakpoints on regular labels. That could be useful during debugging - to catch when a particular "case" was actually hit. 

But you are right. We don't. 
I am still not sure if there is any value in deoptimizing this in Debug case or not.

File: src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs [hunk 3]
```diff
@@ -12,18 +12,21 @@ internal partial struct SwitchIntegralJumpTableEmitter
...
+            // sorted case labels
+            private readonly ImmutableArray<KeyValuePair<ConstantValue, object>> _allLabels;
+
             // range of sorted case labels within this bucket
             private readonly int _startLabelIndex;
             private readonly int _endLabelIndex;
 
-            // sorted case labels
-            private readonly ImmutableArray<KeyValuePair<ConstantValue, object>> _allLabels;
+            private readonly bool _isKnownDegenerate;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs]: >_isKnownDegenerate [](start = 34, length = 18)

Could you please explain what does it mean for the bucket to be "Degenerate"?

File: src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs [hunk 4]
```diff
@@ -60,28 +82,88 @@ internal ulong BucketSize
...
+            // On the other hand, handling a degenerate bucket requires only 1 compare (range check) 
+            // so crumbling one bucket into two degenerate ones, when possible, is advantageous.
+            //
+            // One relatively common case that may result in degenerate switches is in async/iterator 
+            // state machines when control flow must use cascading dispatch into try blocks
+            internal int DegenerateBucketSplit
             {
                 get
                 {
                     if (_startLabelIndex == _endLabelIndex)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs]: >if (_startLabelIndex == _endLabelIndex) [](start = 20, length = 39)

Should _isKnownDegenerate be considered in this function?

File: src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs [hunk 5]
```diff
@@ -60,28 +82,88 @@ internal ulong BucketSize
...
             }
 
+            private bool IsContiguous(ConstantValue lastConst, ConstantValue nextConst)
+            {
+                if (!lastConst.IsNumeric || !nextConst.IsNumeric)
+                {
+                    return false;
+                }
+
+                return nextConst.Int64Value - lastConst.Int64Value == 1;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs]: >return nextConst.Int64Value - lastConst.Int64Value == 1; [](start = 16, length = 56)

The function below sometimes looks at constants as UInt64Value. Should we do something similar here?

File: src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs [hunk 6]
```diff
@@ -60,28 +82,88 @@ internal ulong BucketSize
...
+            // Computed jump costs the same or more expensive than a regular conditional branch.
+            // So compbined cost would be 2+ conditional branches.
+            // Based on benchmarks it is actually closer to 3
+            //
+            // On the other hand, handling a degenerate bucket requires only 1 compare (range check) 
+            // so crumbling one bucket into two degenerate ones, when possible, is advantageous.
+            //
+            // One relatively common case that may result in degenerate switches is in async/iterator 
+            // state machines when control flow must use cascading dispatch into try blocks
+            internal int DegenerateBucketSplit
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs]: >DegenerateBucketSplit [](start = 25, length = 21)

It is not clear to me why we are supporting splitting only in two buckets. Why not in more?

Comment (VSadov) [inline src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs]: The input is a "dense" bucket - the one that previous heuristics determined as not worth splitting, It has rough execution cost of a 1 regular branch (range check) and 1 computed branch (which cost roughly the same or perhaps more).
The only way to surely beat that cost via splitting is if the bucket can be split into 2 degenerate buckets. Then we have 2 regular branches.

3 degenerate buckets would require up to 3 regular compares. On some hardware computed jumps may cost significantly more than regular ones (because they are harder to predict or whatever), so it could still be profitable, but I did not want to guess that.

Basically if we have 3 degenerate buckets that can be merged into a dense bucket, we prefer a dense bucket, which we emit as "switch" opcode.

---
In reply to: [93716855](https://github.com/dotnet/roslyn/pull/16054#discussion_r93716855) [](ancestors = 93716855)

File: src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs (hunk 5 above)
Comment (VSadov) [inline src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs]: Yes, I will just make it call GetBucketSize. 

---
In reply to: [93716277](https://github.com/dotnet/roslyn/pull/16054#discussion_r93716277) [](ancestors = 93716277)

File: src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs (hunk 4 above)
Comment (VSadov) [inline src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs]: It would be the same as checking for buckets of size 1 as only those are known to be degenerate right after construction.
I guess it would be a bit cheaper to check for that

---
In reply to: [93715584](https://github.com/dotnet/roslyn/pull/16054#discussion_r93715584) [](ancestors = 93715584)

File: src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs (hunk 1 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs]: Removed the check for Debug

---
In reply to: [93700623](https://github.com/dotnet/roslyn/pull/16054#discussion_r93700623) [](ancestors = 93700623)

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - I believe all the CR feedback has been addressed.

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - ping.

File: src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs]: Consider capturing this explanation in a comment.

---
In reply to: [93723682](https://github.com/dotnet/roslyn/pull/16054#discussion_r93723682) [](ancestors = 93723682,93716855)

**Comment by AlekseyTs** [comment]
LGTM. However, consider capturing the explanation you gave in https://github.com/dotnet/roslyn/pull/16054#discussion_r93723682 as a comment so that the rational behind the splitting would be clear.

**Comment by VSadov** [comment]
@AlekseyTs - will put the explanation into a source comment.

**Comment by VSadov** [comment]
@srivatsn - May I please have approval for an RC.3 merge? Thanks!

**Comment by AlekseyTs** [comment]
LGTM

**Comment by VSadov** [comment]
Thanks!!
